crates/holochain/src/core/workflow/incoming_dht_ops_workflow.rs

```diff
@@ -89,111 +111,46 @@ async fn should_keep(op: &DhtOp) -> WorkflowResult<bool> {
-    fn flush_to_txn_ref(&mut self, writer: &mut Writer) -> WorkspaceResult<()> {
-        self.validation_limbo.0.flush_to_txn_ref(writer)?;
-        self.element_pending.flush_to_txn_ref(writer)?;
-        self.meta_pending.flush_to_txn_ref(writer)?;
-        self.meta_integrated.flush_to_txn_ref(writer)?;
-        Ok(())
-    }
+fn add_to_pending(
+    env: &EnvWrite,
+    op: DhtOpHashed,
+    from_agent: Option<AgentPubKey>,
+    request_validation_receipt: bool,
+) -> StateMutationResult<()> {
+    let send_receipt = needs_receipt(&op, &from_agent) && request_validation_receipt;
+    tracing::debug!(?op);
+
+    let op_hash = op.as_hash().clone();
+    env.conn()?.with_commit(|txn| {
+        insert_op(txn, op, false)?;
+        set_require_receipt(txn, op_hash, send_receipt)?;
+        StateMutationResult::Ok(())
+    })?;
+    Ok(())
 }
 
-impl IncomingDhtOpsWorkspace {
-    pub fn new(env: EnvironmentRead) -> WorkspaceResult<Self> {
-        let db = env.get_db(&*INTEGRATED_DHT_OPS)?;
-        let integrated_dht_ops = KvBufFresh::new(env.clone(), db);
-
-        let db = env.get_db(&*INTEGRATION_LIMBO)?;
-        let integration_limbo = KvBufFresh::new(env.clone(), db);
-
-        let validation_limbo = ValidationLimboStore::new(env.clone())?;
-
-        let element_pending = ElementBuf::pending(env.clone())?;
-        let meta_pending = MetadataBuf::pending(env.clone())?;
-
-        let meta_integrated = MetadataBuf::vault(env)?;
-
-        Ok(Self {
-            integration_limbo,
-            integrated_dht_ops,
-            validation_limbo,
-            element_pending,
-            meta_pending,
-            meta_integrated,
-        })
-    }
-
-    fn add_to_pending(
-        &mut self,
-        hash: DhtOpHash,
-        op: DhtOp,
-        from_agent: Option<AgentPubKey>,
-        request_validation_receipt: bool,
-    ) -> DhtOpConvertResult<()> {
-        let send_receipt = needs_receipt(&op, &from_agent) && request_validation_receipt;
-        let basis = op.dht_basis();
-        let op_light = op.to_light();
-        tracing::debug!(?op_light);
-
-        // register the highest observed header in an agents chain
-        if let DhtOp::RegisterAgentActivity(_, header) = &op {
-            self.meta_integrated.register_activity_observed(
-                header.author(),
-                HighestObserved {
-                    header_seq: header.header_seq(),
-                    hash: vec![op_light.header_hash().clone()],
-                },
-            )?;
-        }
-
-        integrate_single_data(op, &mut self.element_pending)?;
-        integrate_single_metadata(
-            op_light.clone(),
-            &self.element_pending,
-            &mut self.meta_pending,
+pub fn op_exists(vault: &EnvWrite, hash: &DhtOpHash) -> DatabaseResult<bool> {
+    let exists = vault.conn()?.with_reader(|txn| {
+        let mut stmt = txn.prepare(
+            "
+            SELECT 
+            *
```

Comment:
I don't think this is very efficient for `exists` check, as the underlying query still builds the response data structures.